### PR TITLE
fix: ci pipeline build context

### PR DESCRIPTION
**Pull request by AbhishekA1509**
# Description
Creating a separate component for Build Context and removing its duplicate instances, have also shifted the corresponding states and useEffects into that component only and since its flag based so would also be performance optimisation.

Fixes # (issue)

### CODE REVIEW

File: src/components/ciConfig/BuildContext.tsx [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+        formState.repository.value = selectedBuildContextGitMaterial.name
+        setCurrentCIBuildConfig({
+            ...currentCIBuildConfig,
+            buildContextGitMaterialId: selectedBuildContextGitMaterial.id,
+        })
+    }
+
+    const handleBuildContextCheckoutPathChange = (checkoutPath) => {
+        const val = checkoutPath.value
+        let flag = false
```
Comment (vivek-devtron) [inline src/components/ciConfig/BuildContext.tsx:107]: we can directly write it as `const flag = checkoutPath.value === RootBuildContext`

File: src/components/ciConfig/BuildContext.tsx [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        if (val === RootBuildContext) {
+            flag = true
+        }
+        setUseRootBuildContextFlag(flag)
```
Comment (vivek-devtron) [inline src/components/ciConfig/BuildContext.tsx:111]: Why this is getting set 3 times?

File: src/components/ciConfig/BuildContext.tsx [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+                                classNamePrefix="build-config__select-checkout-path-for-build-context"
+                                tabIndex={4}
+                                isMulti={false}
+                                isClearable={false}
+                                isSearchable={false}
+                                options={checkoutPathOptions}
+                                getOptionLabel={(option) => `${option.label}`}
+                                getOptionValue={(option) => `${option.value}`}
+                                value={getCheckoutPathValue(useRootBuildContextFlag)}
+                                styles={{
```
Comment (vivek-devtron) [inline src/components/ciConfig/BuildContext.tsx:229]: Please extract this out

File: src/components/ciConfig/CIDockerFileConfig.tsx [hunk 4]
```diff
@@ -107,46 +101,30 @@ export default function CIDockerFileConfig({
...
+            const currentCIBuildConfig = ciConfig?.ciPipelines?.[0]?.dockerConfigOverride?.ciBuildConfig
+            
+            const currentOverriddenBuildContext = currentCIBuildConfig?.dockerBuildConfig?.buildContext
+            const currentOverriddenGitMaterialId = currentCIBuildConfig?.gitMaterialId
+            const currentOverriddenBuildContextGitMaterialId = currentCIBuildConfig?.buildContextGitMaterialId
+
+            // TODO: Ask whether second check is redundant since string empty comes in ! case
+            const isSameCurrentBuildContext =
+                currentOverriddenGitMaterialId === currentOverriddenBuildContextGitMaterialId &&
+                (!currentOverriddenBuildContext || currentOverriddenBuildContext === '')
```
Comment (vivek-devtron) [inline src/components/ciConfig/CIDockerFileConfig.tsx:113]: Please remove the `currentOverriddenBuildContext === '` check

File: src/components/ciConfig/CIDockerFileConfig.tsx [hunk 5]
```diff
@@ -107,46 +101,30 @@ export default function CIDockerFileConfig({
...
+
+            const isSameGlobalBuildContext =
+                currentMaterial.id === currentBuildContextGitMaterial.id &&
+                (!ciConfig?.ciBuildConfig?.dockerBuildConfig?.buildContext ||
+                    ciConfig?.ciBuildConfig?.dockerBuildConfig?.buildContext === '')
```
Comment (vivek-devtron) [inline src/components/ciConfig/CIDockerFileConfig.tsx:118]: Same here

Comment (AbhishekA1509) [reply src/components/ciConfig/CIDockerFileConfig.tsx]: Done

File: src/components/ciConfig/CIDockerFileConfig.tsx (hunk 4 above)
Comment (AbhishekA1509) [reply src/components/ciConfig/CIDockerFileConfig.tsx]: Done

File: src/components/ciConfig/BuildContext.tsx (hunk 3 above)
Comment (AbhishekA1509) [reply src/components/ciConfig/BuildContext.tsx]: Done

File: src/components/ciConfig/BuildContext.tsx (hunk 1 above)
Comment (AbhishekA1509) [reply src/components/ciConfig/BuildContext.tsx]: Done

File: src/components/ciConfig/BuildContext.tsx (hunk 2 above)
Comment (AbhishekA1509) [reply src/components/ciConfig/BuildContext.tsx]: Fixed

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**